### PR TITLE
first commit - adding invoices for microbial

### PR DESCRIPTION
**Pull request by mayabrandi**
This PR adds Microbial invoicing to the Invoicing tool

I have just added code following the code layout that already is in place for samples and pools. 
I intend to update the structure a bit, add dokumentation to all functions in the code etc. in a separate PR.

How to test:
0. install on stage:
sh ~/servers/resources/clinical-db.scilifelab.se/update-clinical-api-stage.sh microbial
test it [here](https://clinical-api-stage.scilifelab.se/invoices/new/Microbial)

1. select: New Invoice - Microbial. Only customers with microbial samples should show up in the drop down.
2. select customer. - Only microbial samples should show up
3. select samples. - Should give no errors
4. create invoice. - Should give no errors

- [x] functionality tested by @emiliaol 
- [x] code reviewed by @ingkebil 
- [x] code approved for deploy by @ingkebil 

This is a minor because the changes are backwards compatible.

### CODE REVIEW

**Comment by emiliaol** [comment]
Error found.

Tested on clinical-db stage:
Skärmavbild 2019-06-03 kl  16 32 13

The microbial invoicing is there:
Skärmavbild 2019-06-03 kl  16 29 52

The customers in the list are customers with microbial samples:
Skärmavbild 2019-06-03 kl  16 30 11

The samples loaded are microbial/viral samples:
Skärmavbild 2019-06-03 kl  16 31 19

Creating the invoice leads to a 404 error:
Skärmavbild 2019-06-03 kl  16 31 37

File: cg/server/invoices/views.py [hunk 1]
```diff
@@ -108,6 +117,8 @@ def new(record_type):
         records, customers_to_invoice = db.samples_to_invoice(customer=customer_obj)
     elif record_type=='Pool':
         records, customers_to_invoice = db.pools_to_invoice(customer=customer_obj)
+    if record_type=='Microbial':
```
[the remark this answers is not in the document]
Comment (mayabrandi) [reply cg/server/invoices/views.py]: 👍

File: cg/store/api/status.py [hunk 2]
```diff
@@ -505,6 +505,25 @@ def samples_not_downsampled(self):
         )
         return records
 
+    def microbial_samples_to_invoice(self, customer: models.Customer = None):
+        """Fetch microbial samples that should be invoiced.
+
+        Return microbial samples have been delivered but invoiced, excluding those that
```
[the remark this answers is not in the document]
Comment (mayabrandi) [reply cg/store/api/status.py]: 👍

File: cg/store/api/status.py [hunk 3]
```diff
@@ -505,6 +505,25 @@ def samples_not_downsampled(self):
...
+            self.MicrobialSample.query.filter(
+                models.MicrobialSample.delivered_at != None,
+                models.MicrobialSample.invoice_id == None,
+                #models.MicrobialSample.no_invoice == False,
+                #models.MicrobialSample.downsampled_to == None
+            )
+        )
+        customers_to_invoice = [record.microbial_order.customer for record in records.all()]
+        customers_to_invoice = list(set(customers_to_invoice))
+        records = records.join(models.MicrobialOrder).filter(models.MicrobialOrder.customer_id == customer.id) if customer else records
```
[the remark this answers is not in the document]
Comment (mayabrandi) [reply cg/store/api/status.py]: 👍

File: cg/meta/invoice.py [hunk 4]
```diff
@@ -137,7 +150,7 @@ def get_price(self, discount: int, record: models.Sample):
         """Get discount price for a sample."""
         if type(record)==models.Pool:
             priority='research'
-        elif type(record)==models.Sample:
+        else:
             priority = record.priority_human
```
[the remark this answers is not in the document]
Comment (mayabrandi) [reply cg/meta/invoice.py]: "priority_human" as in human readable priority! You can see how it is set up in store/models.py!

File: cg/meta/invoice.py [hunk 5]
```diff
@@ -70,6 +69,17 @@ def prepare(self, costcenter: str) -> dict:
...
+            record_type = 'Microbiellt Prov'
+            for microbial_samples in self.invoice_obj.microbial_samples:
+                record = self.prepare_record(
+                    costcenter=costcenter.lower(),
+                    discount=self.invoice_obj.discount,
+                    record=microbial_samples)
+                if record:
+                    records.append(record)
+                else:
+                    return None
```
[the remark this answers is not in the document]
Comment (mayabrandi) [reply cg/meta/invoice.py]: Okej. I think I understand what you mean. I can try to do that in the next pr witch is only about prettifying the invoicing code. Thank you.

**Comment by mayabrandi** [comment]
@emiliaol I fixed the bug. Would you mind try it again?

**Comment by emiliaol** [comment]
Tested again and this time I could generate the invoice:

Skärmavbild 2019-06-11 kl  16 30 07

Downloading the invoice template with the button `Get Invoice Template` also worked. See attached excel sheet.

[Invoice_265_KTH.xlsx](https://github.com/Clinical-Genomics/cg/files/3276993/Invoice_265_KTH.xlsx)

I have not tested the other buttons/functionalities as I do not know what to expect from them.

**Comment by mayabrandi** [comment]
Thanks @emiliaol

File: cg/meta/invoice.py [hunk 6]
```diff
@@ -14,15 +14,35 @@ def __init__(self, db: Store, lims_api: lims.LimsAPI, invoice_obj: models.Invoic
         self.log = []
         self.invoice_obj = invoice_obj
         self.customer_obj = invoice_obj.customer
+        self.record_type = ''
+        self.raw_records = []
+        self._get_record_type()
+
+
+    def _get_record_type(self):
```
Comment (ingkebil) [inline cg/meta/invoice.py]: you're not getting anything with this method, e.g. there is no return statement. You are however setting the record_type. Why not name it that way?

File: cg/meta/invoice.py [hunk 7]
```diff
@@ -14,15 +14,35 @@ def __init__(self, db: Store, lims_api: lims.LimsAPI, invoice_obj: models.Invoic
...
+
         msg = f'Could not open/generate invoice. Contact information missing in database for ' \
-            f'customer {self.customer_obj.internal_id}. See log files.'
-        if costcenter.lower() == 'kth':
-            customer = self.db.customer('cust999')
-        else:
-            customer = self.customer_obj
+              f'customer {self.customer_obj.internal_id}. See log files.'
+
+        customer = self.db.customer('cust999') if costcenter.lower() == 'kth' else self.customer_obj
```
Comment (ingkebil) [inline cg/meta/invoice.py]: Something to think about for a later iteration: never hardcode. A possible solution is to add an extra field to the customer table denoting that this customer is special in the way 'cust999' is special. That way you can query it, exclude it from queries, etc...

No need to refactor that now.

File: cg/meta/invoice.py (hunk 6 above)
Comment (ingkebil) [reply cg/meta/invoice.py]: There is no more record_type 'Prov'?

Comment (ingkebil) [reply cg/meta/invoice.py]: oh, I see. You renamed that?

File: cg/meta/invoice.py [hunk 8]
```diff
@@ -32,54 +52,32 @@ def prepare_contact_info(self, costcenter):
...
-                else:
-                    return None
+    
+        for raw_record in self.raw_records:
+            if self.record_type == 'Pool':
+                pooled_samples += genologics_lims.samples_in_pools(raw_record.name, raw_record.ticket_number)
+            record = self.prepare_record(
+                            costcenter=costcenter.lower(),
+                            discount=self.invoice_obj.discount,
+                            record=raw_record)
```
Comment (ingkebil) [inline cg/meta/invoice.py]: :+1:

File: cg/meta/invoice.py [hunk 9]
```diff
@@ -95,34 +93,31 @@ def prepare(self, costcenter: str) -> dict:
...
-            discounted_price = self.get_price(discount, record)
-        except:
-            self.log.append(f'Application tag/version semms to be missing for sample {record.id}.')
+
+    def _discount_price(self, record, discount: int = 0):
+        """Get discount price for a sample, pool or microbial_sample."""
+
+        priority = 'research' if self.record_type == 'Pool' else record.priority_human
+
+        full_price = getattr(record.application_version, f"price_{priority}")
```
Comment (ingkebil) [inline cg/meta/invoice.py]: Clean! :clap:

File: cg/meta/invoice.py [hunk 10]
```diff
@@ -95,34 +93,31 @@ def prepare(self, costcenter: str) -> dict:
...
+        """Split price based on cost center"""
+        if price:
             try:
                 if costcenter == 'kth':
                     split_factor = percent_kth / 100
                 else:
                     split_factor = (100 - percent_kth) / 100
-                price = round(discounted_price * split_factor, 1)
+                split_price = round(price * split_factor, 1)
             except:
```
Comment (ingkebil) [inline cg/meta/invoice.py]: Why would this ever fail?

File: cg/meta/invoice.py [hunk 11]
```diff
@@ -131,9 +126,27 @@ def prepare_record(self, costcenter: str, discount: int, record: models.Sample):
...
+    def prepare_record(self, costcenter: str, discount: int, record):
+        """Get invoice information for a specific sample, pool or microbial_sample."""
+
+        try:
+            tag = record.application_version.application.tag
+            version = str(record.application_version.version)
+            percent_kth = record.application_version.application.percent_kth
+            discounted_price = self._discount_price(record, discount)
+        except:
+            self.log.append(f'Application tag/version semms to be missing for sample {record.id}.')
```
Comment (ingkebil) [inline cg/meta/invoice.py]: semms -> seems

**Comment by ingkebil** [comment]
I also noticed that you have a ton of linting errors that need fixing. Please check the TravisCI at the bottom of the log (you'll have to scroll all the way down, click on "view raw log", and scroll all the way down again.

**Comment by barrystokman** [comment]
> Non trivial linting errors to fix

>Linting file: [1mcg/store/api/status.py[0m
line 515, col 16: Convention: [singleton-comparison]: StatusHandler.microbial_samples_to_invoice: Comparison to None should be 'expr is not None'
line 515, col 53: [E711]: comparison to None should be 'if cond is not None:'
line 516, col 16: Convention: [singleton-comparison]: StatusHandler.microbial_samples_to_invoice: Comparison to None should be 'expr is None'
line 517, col 0: Convention: [line-too-long]: Line too long (105/100)
line 517, col 17: [E265]: block comment should start with '# '
line 517, col 101: [E501]: line too long (105 > 100 characters)
line 518, col 0: Convention: [line-too-long]: Line too long (107/100)
line 518, col 17: [E265]: block comment should start with '# '
line 518, col 101: [E501]: line too long (107 > 100 characters)
line 524, col 0: Convention: [line-too-long]: Line too long (114/100)
line 524, col 101: [E501]: line too long (114 > 100 characters)

@mayabrandi I see that testing and review is done, except for some leftover linting errors (see above). I think if those are fixed @ingkebil might approve for deployment soon.

**Comment by mayabrandi** [comment]
I fixed them, but new came up that I don’t know how to fix. There must be another way to deal with these ”errors”. So one at least get all at once. Or I just don’t know how to work with Linting in a efficient way. Maybe you could give me a crash course?

**Comment by barrystokman** [comment]
@mayabrandi I fixed the bare exceptions by using the a generic `ValueError` from the built-in exceptions, since it seems the exception handling tries to handle values not being found. Can you check if that makes sense? Feel free to change it into something more fitting if you want.

**Comment by ingkebil** [comment]
My two cents: there is no easy way of going through these linting errors. We can speed up the process by running `git lint` locally on a cloned repo reset to the first commit of this branch. ... haha ... I hope that sentence made sense ;)

From the top of my head, that would be something along these lines:
```
git clone -b microbial https://github.com/Clinical-Genomics/cg.git
cd cg
git reset --soft 0131c20d662ec7800065c3c92110b07cc6f628e5
git lint
```

This would lint only your changes and not the whole of `cg`.

The biggest challenge comes from how to test your changes, as you can only use this clone for inspection and not for fixing.
I can show you once I'm back.

A better way would be to have an editor that would show you linting errors: pycharm, visual studio code, and vim (with syntastic plugin) all do this.

But now that we're here, maybe someone has a better idea?

**Comment by mayabrandi** [comment]
> A better way would be to have an editor that would show you linting errors: pycharm, visual studio code, and vim (with syntastic plugin) all do this.
> 
> 


This is what I want. Should I select pylint??

**Comment by ingkebil** [comment]
Yes, go for it :)

Pylint and line length set to 100 instead of pylint's default of 78.

**Comment by barrystokman** [comment]
@ingkebil With the linting fixed, is this ready to be approved for deployment?

**Comment by ingkebil** [comment]
Yes :)

**Comment by ingkebil** [comment]
I'll approve when the version bump is motivated :)

**Comment by barrystokman** [comment]
> I'll approve when the version bump is motivated :)

@mayabrandi can you add this so we can deploy?

**Comment by ingkebil** [comment]
Feel free to :metal:

**Pull request merged**